### DISCUSSION TITLE
Solidity合约常用技巧

### DISCUSSION THREAD

**Discussion by lispking** [openbuildxyz/Web3-Frontend-Bootcamp#1013] [General]
# 背景

由于[工程问题](https://github.com/openbuildxyz/Web3-Frontend-Bootcamp/discussions/1011)引起了激烈讨论，在这里慢慢沉淀一些好用的工具以及Solidity合约的常用技巧，逐步更新生产级用到的技巧，这些知识都是价值X亿，请大家惠存

## 那些好玩的入门教程

* 僵尸教程: https://cryptozombies.io/

### 部署和测试合约必备利器（二选一），没用过`remix`，但是生产环境很少会用

* hardhat: https://hardhat.org/hardhat-runner/docs/getting-started#overview
* getfoundry https://book.getfoundry.sh/

## 那些常用的库

* openzeppelin（必熟）: https://docs.openzeppelin.com/

## 那些有用的写法

* ERC20（必学）：https://docs.openzeppelin.com/contracts/5.x/api/token/erc20
* ERC721（必学）：https://docs.openzeppelin.com/contracts/5.x/api/token/erc721
* 合约升级（必学）：https://docs.openzeppelin.com/upgrades
* 权限访问（必学）：https://docs.openzeppelin.com/contracts/5.x/api/access

## 那些常用的库

* MerkleProof（一般用于空投校验）：https://docs.openzeppelin.com/contracts/5.x/api/utils#MerkleProof
* ReentrancyGuard（重入攻击）：https://docs.openzeppelin.com/contracts/5.x/api/utils#ReentrancyGuard
* Pausable（合约暂停，非常有用，防止合约出问题可暂停服务）：https://docs.openzeppelin.com/contracts/5.x/api/utils#Pausable
* Math（数学计算用的）：https://docs.openzeppelin.com/contracts/5.x/api/utils#Math
* EnumerableMap（最近一个亿级项目用到，非常好用）: https://docs.openzeppelin.com/contracts/5.x/api/utils#EnumerableMap
* EnumerableSet（最近一个亿级项目用到，非常好用）: https://docs.openzeppelin.com/contracts/5.x/api/utils#EnumerableSet

## 那些深入的知识

* Web3.0架构（相当于本课程的总结）：https://www.preethikasireddy.com/post/the-architecture-of-a-web-3-0-application
* 合约全貌（必读）：https://docs.soliditylang.org/en/v0.8.26/contracts.html
* ABI规范（必读）：https://docs.soliditylang.org/en/v0.8.26/abi-spec.html
* 合约调试（万能的打印）：https://hardhat.org/tutorial/debugging-with-hardhat-network#solidity-console-log
* 合约测试（UT）：https://hardhat.org/tutorial/testing-contracts
* Gas源由：https://ethereum.org/zh/developers/docs/gas/
* EVM介绍：https://ethereum.org/zh/developers/docs/evm/
* EVM指令集解密：https://takenobu-hs.github.io/downloads/ethereum_evm_illustrated.pdf
* 共识算法：https://ethereum.org/zh/developers/docs/consensus-mechanisms/
* 理解女巫攻击：https://en.wikipedia.org/wiki/Sybil_attack
* Oracle（预言机）：https://ethereum.org/zh/developers/docs/oracles/

## 持续更新中...

**Comment by xiangnuans** [comment]
# 作为智能合约入门者（前端开发背景），聊一下智能合约学习过程和选择：

其实在学习做任务之前去了解了一圈，工具和框架
对比之后选用的hardhat，也尝试了用remix、truffle
- remix 打开体验了下官方提供的Demo，浏览器编译插件，习惯了用客户端开发工具，快捷键操作，所以这点在浏览器上使用Remix体验不是太好，所以放弃了；
- 使用truffle也初始化项目，踩了好几次坑，应该是刚入门，很多不熟悉
- 最后选择了hardhat，作为前端开发者基本无任何障碍，使用体验顺滑。

# 再回过头看

- 项目还是得用上述大佬推荐的两种工具；
- Remix的优势，我觉得好的使用方式是，课程里讲师推荐的，可以使用remixd，代替了写编译脚本，但是如果是业务或者自己想测试，比如task3 执行购买NFT或者上下架等等，还是编写脚本感觉更合适一些

**Reply by ghost** [reply]
你好，我想问一下，刚刚从web2转型，但是找工作他要web3工作经验。这种情况下，要怎么办呀

**Reply by lispking** [reply]
OpenBuild有开源项目可贡献，建议先积累经验，再找会容易些
